Update installation instruction to fit recent CUDA11 and torch 1.8+ push (#969)
ivadomed/ivadomed: docs/source/installation.rst

============

Supported OS
++++++++++++

Currently, ``ivadomed`` only supports GPU/CPU on ``Linux`` and ``Windows`` and CPU only on ``macOS`` and `Windows Subsystem for Linux <https://docs.microsoft.com/en-us/windows/wsl/>`_.

Step 1: Setup dedicated python environment
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

You can setup ``ivadomed`` using either Conda or Venv:

.. tabs::

    .. tab:: Install via ``conda``

        This is the easiest way for personal computers.

        1. Create new conda environment using ``environment.yml`` file

        ::

>             conda env create --file environment.yml

Reviewer: Now that I think about it, I'm a bit confused about the purpose of this `environment.yml` file. 

https://github.com/ivadomed/ivadomed/blob/c226c49c384a8c322b99f9529f10b408b6687637/environment.yml#L1-L9

Since all of the dependencies are installed via `pip` + `requirements.txt`, it seems redundant with Step 3 of this page, where `requirements.txt` will be installed once more.

(Also, I notice the name is still hardcoded as `IvadoMedEnv` in this file, so it would probably need to be updated to `ivadomed_env` as well.)

Author: Good point. I think the original rationale was to actually have a one stop installation for all dependencies including torch etc (as those distribute via conda rather well) but then the issue with that is, GPU are not always supported and also we won't know at installation time which CUDA to use etc, which significantly complicate the matter. 

I like your suggestion to keep it simple so I will just keep the steps simple here to create an empty environment and leave the rest alone (removing pip installation part)

        2. Activate the new conda environment (default named ``ivadomed_env``)

        ::

            conda activate ivadomed_env

    .. tab:: Install via ``venv``

        1. Setup Python Venv Virtual Environment.

        ``ivadomed`` requires Python >= 3.6 and <3.9 (If you are using `Compute Canada <https://www.computecanada.ca/>`_, you can load modules (e.g. python 3.9) as `mentioned here <https://intranet.neuro.polymtl.ca/computing-resources/compute-canada#modules>`_ and `also here <https://docs.computecanada.ca/wiki/Utiliser_des_modules/en#Loading_modules_automatically>`_ ). We recommend

>         working under a virtual environment, which could be set as follows:

Reviewer: I think this "Compute Canada" bracketed text might be better off in the Compute Canada tab.

Author: Good point. They were originally sort of together. Now separated. Moving there.

        ::

            python -m venv ivadomed_env

        2. Activate the new virtual environment (default named `ivadomed_env`)

        ::

            source ivadomed_env/bin/activate

        .. warning::
           If the default Python version installed in your system does not fit the version requirements, you might need to specify a version of Python associated with your virtual environment:

           ::

>              python -m venv ivadomed_env --python=python3.7

Reviewer: I've never seen this `-python=` syntax... and when I try it, I get an error:

```
joshua@tadpole:~$ python -m venv ivadomed_env --python=python3.7
usage: venv [-h] [--system-site-packages] [--symlinks | --copies] [--clear]
            [--upgrade] [--without-pip] [--prompt PROMPT] [--upgrade-deps]
            ENV_DIR [ENV_DIR ...]
venv: error: unrecognized arguments: --python=python3.7
```

Just curious, where did you find this? o:

Author: Oh man... VENV doesn't support this :(. Okay. Will have to remove this section then and hope they use the right python version to begin with.

Reviewer: Thankfully, the range of versions that ivadomed supports is quite wide, so I think in the vast majority of cases, the default Python will be OK.

Reviewer: Ah, this has a better solution in https://github.com/ivadomed/ivadomed/pull/969#discussion_r744307483.

    .. tab:: Compute Canada HPC

        There are numerous constraints and limited package availabilities with ComputeCanada cluster environment.

        It is best to attempt ``venv`` based installations and follow up with ComputeCanada technicall support as MANY specially compiled packages (e.g. numpy) are exclusively available for Compute Canada HPC environment.

Step 2: Install ``ivadomed``
++++++++++++++++++++++++++++

.. tabs::

    .. tab:: Pypi Installation

        Install ``ivadomed`` and its requirements from

>         `Pypi <https://pypi.org/project/ivadomed/>`__:

Reviewer: Nitpick: `Pypi` -> [`PyPI`](https://pypi.org/)

Author: Nice. TIL. Totally didn't know that before.

        Bleeding-edge developments are available on the project's master branch
        on Github. Installation ``ivadomed`` from source:

        ::

            git clone https://github.com/ivadomed/ivadomed.git

            cd ivadomed

            pip install -e .

Step 3: Install ``ivadomed`` with CPU or GPU Support, Install ``torch`` and ``torchvision``
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

.. tabs::

    .. tab:: CPU Support

        If you plan to run ``ivadomed`` on CPU only, run the following command:
        ::

            pip install -r requirements.txt

    .. tab:: Nvidia GPU Support

        ``ivadomed`` requires CUDA11 to execute properly. If you have a nvidia GPU, try to look up its Cuda Compute Score `here <https://developer.nvidia.com/cuda-gpus>`_, which needs to be > 3.5 to support CUDA11. Then, make sure to upgrade to nvidia driver to be at least v450+ or newer.

        If you have a compatible NVIDIA GPU that supports CUDA11 and with the right driver installed, try run the following command:

        ::

           pip install -r requirements_gpu.txt

>         Please note that this must happens after the previous ``ivadomed`` installation step.

Reviewer: I think that this line might be a little unnecessary, just because I think it's a fair expectation that people should follow steps in order, hehe.

Author: Good point. I was actually thinking about this too.

Additional Steps Required For ``ivadomed`` Developers: Additional Dependencies and ``pre-commit``

> +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

Reviewer: This is a nitpicky thing, but what if we had the headers for this section be:

```rest
Developer-only installation steps
+++++++++++++++++++++++++++++++++

The additional steps below are only necessary for contributors to the ``ivadomed`` project.

Additional dependencies and ``pre-commit``
------------------------------------------

Blah blah blah...

``macOS``-specific development steps
------------------------------------

Blah blah blah...
```

That way, the titles can be a little shorter/easier to read, especially in the table-of-contents.

![Screenshot from 2021-11-05 12-25-30](https://user-images.githubusercontent.com/16181459/140544587-5a774cec-b232-4172-a8e4-0997adcf8131.png)

Author: Love it! I just couldn't recall the Restructure text heading hierarchy.

Reviewer: > Love it! I just couldn't recall the Restructure text heading hierarchy.

The interesting thing is that RST doesn't actually have a heading hierarchy! You can use the symbols in any order, and `docutils`/`sphinx` will interpret each new symbol as the next-smallest header. ([source](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#sections))

> Normally, there are no heading levels assigned to certain characters as the structure is determined from the succession of headings.

This can get a bit weird, though, because header symbols will often end up inconsistent between pages. :sweat_smile:

Author: Thanks for the in-depth explanation.

============

Supported OS
++++++++++++

Currently, ``ivadomed`` only supports GPU/CPU on ``Linux`` and ``Windows`` and CPU only on ``macOS`` and `Windows Subsystem for Linux <https://docs.microsoft.com/en-us/windows/wsl/>`_.

Step 1: Setup dedicated python environment
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

You can setup ``ivadomed`` using either Conda or Venv:

.. tabs::

    .. tab:: Install via ``conda``

>         This is the easiest way for personal computers.

Reviewer: Hmm... I'm wondering, is `conda` really the "easiest way for personal computers"?

Most people won't have conda installed on their machines by default, while a lot of people (e.g. Linux users) will have access to a system Python which they can use to create a virtual environment.

So, for the order of the tabs, I'm wondering if `venv` should be the default tab, while `conda` should be the second tab.

Author: Very good point. `conda` was useful for GPU related installation but so far, guide has not really leveraged that much. Switching to `venv`.

        1. Create new conda environment using ``environment.yml`` file

        ::

            conda env create --name ivadomed_env

        2. Activate the new conda environment (default named ``ivadomed_env``)

        ::

            conda activate ivadomed_env

    .. tab:: Install via ``venv``

        1. Setup Python Venv Virtual Environment.

        ``ivadomed`` requires Python >= 3.6 and <3.10. We recommend working under a virtual environment, which could be set as follows:

        ::

>             python -m venv ivadomed_env

Reviewer 2: This did not work for me, I had to `sudo apt install python3.8-venv` first, and then `python3.8 -m venv ivadomed_env`.

Reviewer 2: Also, there was a previous discussion about being unable to set the python is `venv`. I think this is problematic, and as a user, I want to be able to choose my version. I think the way to do this is maybe with:
`python3.X -m venv ivadomed_env`

Reviewer: Ah! Very good point! `python` runs the risk of mapping to `python2`, but if we tell the user to be explicit about their python version, then we avoid this issue.

We could say something like:

```rest
First, make sure that a compatible version of Python 3 is installed on your system by running:

.. code::

   python3 --version

If your system's Python is not 3.7, 3.8, or 3.9 (or if you don't have Python 3 installed at all), please [install Python](https://realpython.com/installing-python/) before continuing.

Once you have a supported version of Python installed, run the following command (where `python3.X` corresponds to the installed Python version)

.. code::

   python3.X -m venv ivadomed_env
```

Author: Very good point. I didn't consider this. Added. :)

>         2. Activate the new virtual environment (default named `ivadomed_env`)

Reviewer: ```suggestion
        2. Activate the new virtual environment (default named ``ivadomed_env``)
```

(Needs a second set of ` to make it code-formatted.)

        ::

            source ivadomed_env/bin/activate

    .. tab:: Compute Canada HPC

        There are numerous constraints and limited package availabilities with ComputeCanada cluster environment.

>         It is best to attempt ``venv`` based installations and follow up with ComputeCanada technicall support as MANY specially compiled packages (e.g. numpy) are exclusively available for Compute Canada HPC environment.

Reviewer: Typo: `technicall` -> `technical`

>         If you are using `Compute Canada <https://www.computecanada.ca/>`_, you can load modules as `mentioned here <https://intranet.neuro.polymtl.ca/computing-resources/compute-canada#modules>`_ and `also here <https://docs.computecanada.ca/wiki/Utiliser_des_modules/en#Loading_modules_automatically>`_ .

Reviewer: Tiny typo: There's an extra space between `also here` and the `.`.

Step 2: Install ``ivadomed``
++++++++++++++++++++++++++++

.. tabs::

>     .. tab:: Pypi Installation

Reviewer: Nitpick: This still reads as `Pypi`.

        Install ``ivadomed`` and its requirements from
        `PyPI <https://pypi.org/project/ivadomed/>`__:

        ::

            pip install --upgrade pip

            pip install ivadomed

    .. tab:: Repo Installation (Advanced or Developer)

        Bleeding-edge developments are available on the project's master branch

>         on Github. Installation ``ivadomed`` from source:

Reviewer 2: ```suggestion
        on Github. Install ``ivadomed`` from source:
```

        ::

            git clone https://github.com/ivadomed/ivadomed.git

            cd ivadomed

>             pip install -e .

Reviewer 2: I got this error when installing in `venv`:
`ERROR: packaging 21.2 has requirement pyparsing<3,>=2.0.2, but you'll have pyparsing 3.0.5 which is incompatible.`

This prevent `ivadomed` from running. When running CLI I get:
```
Traceback (most recent call last):
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 583, in _build_master
    ws.require(__requires__)
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 900, in require
    needed = self.resolve(parse_requirements(requirements))
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (pyparsing 3.0.5 (/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages), Requirement.parse('pyparsing<3,>=2.0.2'), {'packaging'})

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/mhbourget/ivadomed_env_969_2/bin/ivadomed", line 6, in <module>
    from pkg_resources import load_entry_point
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3252, in <module>
    def _initialize_master_working_set():
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3235, in _call_aside
    f(*args, **kwargs)
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 3264, in _initialize_master_working_set
    working_set = WorkingSet._build_master()
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 585, in _build_master
    return cls._build_from_requirements(__requires__)
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 598, in _build_from_requirements
    dists = ws.resolve(reqs, Environment())
  File "/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages/pkg_resources/__init__.py", line 791, in resolve
    raise VersionConflict(dist, req).with_context(dependent_req)
pkg_resources.ContextualVersionConflict: (pyparsing 3.0.5 (/home/mhbourget/ivadomed_env_969_2/lib/python3.8/site-packages), Requirement.parse('pyparsing<3,>=2.0.2'), {'packaging'})
```

Author: This issue is weirdly related to venv/pip common dependency version which several fixes are available (simplest is to update pip to 21). Could you check if you have Pip v20? In depth details as well as other many useful solutions are reported by @kousu here: https://github.com/neuropoly/computers/issues/218.

We will probably update the dependencies to be more specific to globally address this odd problem.

Reviewer 2: You are right, I had `pip 20.0.2`.
I did `pip install --upgrade pip` and I now have `pip 21.3.1` and the installation and CLI run without error.

Author: Thanks to Nick, we have also added that pyparsing as a requirement and merged to master to include additional details. See #987

> Step 3: Install ``ivadomed`` with CPU or GPU Support, Install ``torch`` and ``torchvision``

Reviewer 2: ```suggestion
Step 3: Install ``torch`` and ``torchvision`` with CPU or GPU Support. 
```

I'm not sure about this, but if I understand correctly "ivadomed" is already instaled at step 2 and step 3 is only for torch and torchvision. Feel free to disregard if I'm mistaken.
The length of the title underline below also has to be adjusted accordingly.

Author: Very good point. I agree the title could be more clear. Changed. :D

++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

.. tabs::

    .. tab:: CPU Support

        If you plan to run ``ivadomed`` on CPU only, run the following command:
        ::

            pip install -r requirements.txt

    .. tab:: Nvidia GPU Support

        ``ivadomed`` requires CUDA11 to execute properly. If you have a nvidia GPU, try to look up its Cuda Compute Score `here <https://developer.nvidia.com/cuda-gpus>`_, which needs to be > 3.5 to support CUDA11. Then, make sure to upgrade to nvidia driver to be at least v450+ or newer.

        If you have a compatible NVIDIA GPU that supports CUDA11 and with the right driver installed, try run the following command:

        ::

           pip install -r requirements_gpu.txt

Developer-only Installation Steps
+++++++++++++++++++++++++++++++++

The additional steps below are only necessary for contributors to the ``ivadomed`` project.

Additional Dependencies and ``pre-commit``
-------------------------------------------

We use ``pre-commit`` to enforce a limit on file size.

> After you've installed ``ivadomed``, install the hooks:

Reviewer: Before, when I suggested the two headings (`+++` and `---`), it was when we had two "developer-only" subsections. But, since the HDF5 warning is removed here, and we have only 1 subsection, then we probably don't need the two different headings anymore. :sweat_smile: 

We could probably just reduce this down to:

```rest
Developer-only dependencies
+++++++++++++++++++++++++++

For contributors to the ``ivadomed`` project, the ``pre-commit`` package is used to enforce a size limit on committed files. 

After you've installed ``ivadomed``, install the ``pre-commit`` hooks by running:

::

>     pip install -r requirements_dev.txt

Reviewer 2: I get this error when installing:
`ERROR: sphinx-tabs 3.2.0 has requirement docutils~=0.16.0, but you'll have docutils 0.17.1 which is incompatible.`

I'm not sure what the impact is, I was able to build the doc locally without errors.

Reviewer: Interesting! It looks like `sphinx-tabs` only updated its requirements (`0.16 -> 0.17`) last month (https://github.com/executablebooks/sphinx-tabs/pull/144), so I'm guessing this hasn't made its way to a release yet.

That said, `pip` should be smart enough to resolve dependencies between all of the requirements in `requirements-dev.txt`. I'm surprised it's throwing an error here. :thinking:

============

Supported OS
++++++++++++

    Currently, ``ivadomed`` supports GPU/CPU on ``Linux`` and ``Windows``, and CPU only on ``macOS`` and `Windows Subsystem for Linux <https://docs.microsoft.com/en-us/windows/wsl/>`_.

Step 1: Setup dedicated python environment
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

    You can setup ``ivadomed`` using either Conda or Venv:

    .. tabs::

        .. tab:: Install via ``venv``

            1. Setup Python Venv Virtual Environment.

                ``ivadomed`` requires Python >= 3.6 and <3.10.

                First, make sure that a compatible version of Python 3 is installed on your system by running:

                .. code::

                   python3 --version

>                 If your system's Python is not 3.7, 3.8, or 3.9 (or if you don't have Python 3 installed at all), please `install Python <https://realpython.com/installing-python/>`_ before continuing.

Reviewer: Ack, sorry. RealPython is now prompting people to create accounts to read their pages.

![Screenshot from 2021-11-13 10-56-55](https://user-images.githubusercontent.com/16181459/141650466-67bc442c-230c-4c03-aeaa-754e8e7d0d8f.png)

Normally I like linking to RealPython because I find their pages very well-written, but this is enough of a nuisance that I think we should maybe link somewhere else.

```suggestion
                If your system's Python is not 3.7, 3.8, or 3.9 (or if you don't have Python 3 installed at all), please `install Python <https://wiki.python.org/moin/BeginnersGuide/Download/>`_ before continuing.
```

This new link is tied to the official Python wiki, so it should be less susceptible to that kind of thing.

Author: Funny I think maybe you visited the site wayyy too often? Just tested it in incognito and it was working? Did they cookie trap you lol. But I agree probably more legit linking the original repo. Going to make the change to official wiki.

>                 If your system's Python is not 3.7, 3.8, or 3.9 (or if you don't have Python 3 installed at all), please `install Python <https://realpython.com/installing-python/>`_ before continuing.

Reviewer 2: ```suggestion
                If your system's Python is not 3.6, 3.7, 3.8, or 3.9 (or if you don't have Python 3 installed at all), please `install Python <https://realpython.com/installing-python/>`_ before continuing.
```

Author: Oh yeah, good point. Not sure why we dropped 3.6 support, even though we technically still test them.

                Once you have a supported version of Python installed, run the following command (replacing ``3.X`` with the Python version number that you installed):

                .. code::

>                    python3.X -m venv ivadomed_env

Reviewer 2: This works great now.
I wanted to mention that the first time I tried this, I had to install python3.8-venv separately (I already had python 3.8 without venv apparently), see this https://github.com/ivadomed/ivadomed/pull/969#discussion_r744296259.

So I don't know if that is normal or not vut a user may have to do the same (I don't know how to test this now that my python3.7-venv is already installed)

Reviewer: > (I already had python 3.8 without venv apparently)

Huh. It was my impression that `venv` came [bundled with Python](https://packaging.python.org/guides/installing-using-pip-and-virtual-environments/#installing-virtualenv):

> Note: If you are using Python 3.3 or newer, the venv module is the preferred way to create and manage virtual environments. venv is included in the Python standard library and requires no additional installation.

But, as you've pointed out, `python-venv` needs to be installed separately on Debian/Ubuntu systems, or else the following error is given:

```
The virtual environment was not created successfully because ensurepip is not
available.  On Debian/Ubuntu systems, you need to install the python3-venv
package using the following command.

    apt install python3.8-venv

You may need to use sudo with that command.  After installing the python3-venv
package, recreate your virtual environment.
```

Yikes. I didn't know that, so thank you for pointing it out. (I missed it the first time you pointed it out, oops.) :sweat:

Reviewer: Apparently this is a bug in Debian/Ubuntu specifically: https://bugs.launchpad.net/ubuntu/+source/python3.4/+bug/1290847

We might benefit from adding an additional message in this section that says:

> If you use Debian or Ubuntu, make sure to install the `python3-venv` module alongside whichever version of Python you choose to install. For other operating systems, `venv` will be installed by default.

But, at the same time, the error message given by Python is quite clear, and provides a command for the user to run, so perhaps we can just leave it be?

Reviewer 2: I like the additional message to help inexperienced users, but I would also be OK to leave it be because the python message is indeed very clear 🙂
In any case, this is minor and I don't think it should block the merge of the PR.

Author: Let's be a bit more verbose. Tab already hide tons of parallel path already for those not interested so our main path is actually already much shorter than others. Better safe than sorry. Thanks for testing this out!

            2. Activate the new virtual environment (default named ``ivadomed_env``)

                ::

                    source ivadomed_env/bin/activate

        .. tab:: Install via ``conda``

            1. Create new conda environment using ``environment.yml`` file

                ::

                    conda env create --name ivadomed_env

>             2. Activate the new conda environment (default named ``ivadomed_env``)

Reviewer: Tiny nitpick: The text "(default named ``ivadomed_env``)" might be a bit unnecessary, since the name ``ivadomed_env`` was chosen in the step immediately before.

            If you are using `Compute Canada <https://www.computecanada.ca/>`_, you can load modules as `mentioned here <https://intranet.neuro.polymtl.ca/computing-resources/compute-canada#modules>`_ and `also here <https://docs.computecanada.ca/wiki/Utiliser_des_modules/en#Loading_modules_automatically>`_.

Step 2: Install ``ivadomed``
++++++++++++++++++++++++++++

    .. tabs::

        .. tab:: PyPI Installation

            Install ``ivadomed`` and its requirements from
            `PyPI <https://pypi.org/project/ivadomed/>`__:

            ::

                pip install --upgrade pip

                pip install ivadomed

        .. tab:: Repo Installation (Advanced or Developer)

            Bleeding-edge developments are available on the project's master branch
            on Github. Install ``ivadomed`` from source:

            ::

                git clone https://github.com/ivadomed/ivadomed.git

                cd ivadomed

                pip install -e .

Step 3: Install ``torch`` and ``torchvision`` with CPU or GPU Support
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

    .. tabs::

        .. tab:: CPU Support

>             If you plan to run ``ivadomed`` on CPU only, install pytorch per instructions `provided here for 1.8 <https://pytorch.org/get-started/previous-versions/#wheel>`_.

Reviewer 2: I find this line a bit confusing.

If I understand correctly:
- if installed with PyPI, torch must be installed from the external link
- if the repo is cloned, it can be installed using requirements

If that is the case, it would need to be more explicitly explained.
Also, linking to external site for instructions can be confusing for user, so I would suggest listing the command to run directly in our instructions.

Reviewer: Same thought! (https://github.com/ivadomed/ivadomed/pull/969#discussion_r748739159) :smile:

Author: Good point! Adopted in conjunction with Joshua's comment. Explicitly stated now for all scenarios.

            Alternatively, if you have already downloaded/cloned the repo with access to the ``requirement.txt`` file, run the following command:
            ::

                pip install -r requirements.txt

>         .. tab:: Nvidia GPU Support

Reviewer: > Alternatively, if you have already downloaded/cloned the repo with access to the ``requirement.txt`` file, run the following command:

Hmm... so, if the user is installing `ivadomed` using `pip install ivadomed` (which is the most prominent method in these new docs), then they won't have access to the requirements files, right? 

Because of that, I feel like the user might get stuck, especially for the GPU tab ("Where is `requirements_gpu.txt`? What do I install?"). And, it feels like a pretty big ask to expect the user clone the repo just to get access to `requirements_gpu.txt`. It kind of defeats the purpose of PyPI, aha. 

But, at the same time, I don't have a great solution to this, because the root cause is deeper than just documentation (https://github.com/ivadomed/ivadomed/issues/861).

Reviewer: As a short-term fix for the docs, maybe we could explicitly list the `torch`/`torchvision` requirements from the requirements files:

CPU Only, Mac:

```rst
.. code::

   pip install torch==1.8.0 torchvision==0.9.0 --find-links https://download.pytorch.org/whl/torch_stable.html
```

CPU-only, Windows/Linux:

```rst
.. code::

   pip install torch==1.8.0+cpu torchvision==0.9.0+cpu --find-links https://download.pytorch.org/whl/torch_stable.html
```

GPU:

```rst
.. code::

   pip install torch==1.8.1+cu111 torchvision==0.9.1+cu111 --find-links https://download.pytorch.org/whl/torch_stable.html
```

This doesn't feel like a great solution, because we're duplicating information from the requirements files. But, addressing #861 probably deserves much more thought/discussion, and I understand that your plan was to finish this PR by today. :thinking:

Author: Good point. I mean, I feel like it almost feel better just to point user to the relevant section in the PyTorch installation guide than trying to duplicate information from them, even though technically the Torch website present these information more cleanly. I have adopted the above into a slightly more elaborate form including original solution.

            ``ivadomed`` requires CUDA11 to execute properly. If you have a nvidia GPU, try to look up its Cuda Compute Score `here <https://developer.nvidia.com/cuda-gpus>`_, which needs to be > 3.5 to support CUDA11. Then, make sure to upgrade to nvidia driver to be at least v450+ or newer.

            If you have a compatible NVIDIA GPU that supports CUDA11 and with the right driver installed, try run the following command:

            ::

               pip install -r requirements_gpu.txt

Developer-only Installation Steps
+++++++++++++++++++++++++++++++++

    The additional steps below are only necessary for contributors to the ``ivadomed`` project.

>     For contributors to the ``ivadomed`` project, the ``pre-commit`` package is used to enforce a size limit on committed files. ``requirements_dev.txt`` also contain additional dependneices related to documentation building etc.

Reviewer 2: ```suggestion
    For contributors to the ``ivadomed`` project, the ``pre-commit`` package is used to enforce a size limit on committed files. ``requirements_dev.txt`` also contain additional dependencies related to documentation building and testing.
```

                .. tabs::

                    .. tab:: Mac/Linux

                        .. code::

                            python3 --version

                    .. tab:: Windows

                        .. code::

                            python --version

                If your system's Python is not 3.6, 3.7, 3.8, or 3.9 (or if you don't have Python 3 installed at all), please `install Python <https://wiki.python.org/moin/BeginnersGuide/Download/>`_ before continuing.

                Once you have a supported version of Python installed, run the following command:

                .. tabs::

                    .. tab:: Mac/Linux

                        .. warning::

                           If you use ``Debian`` or ``Ubuntu``, you may be prompted to install the ``python3-venv`` module when creating the virtual environment. This is expected, so please follow the instructions provided by Python. For other operating systems, ``venv`` will be installed by default.

                        Replacing ``3.X`` with the Python version number that you installed):

                        .. code::

>                             python3.X -m venv ivadomed_env

Reviewer: Nitpick: I'd maybe move some of the text around so that the command comes first:

```rst
                        .. code::

                           # Note: Replace ``3.X`` with the specific Python version number that you installed
                           python3.X -m venv ivadomed_env

                        .. warning::

                           If you use ``Debian`` or ``Ubuntu``, you may be prompted to install the ``python3-venv`` module when creating the virtual environment. This is expected, so please follow the instructions provided by Python. For other operating systems, ``venv`` will be installed by default.
```

(I suggest this because the preceding text is "run the following command:". I found it a bit jarring to see a warning, rather than the actual command. So, I think this way might flow a bit better?)

Reviewer 2: I agree with Joshua here. I would also put the command first, followed by a note about Debian/Ubuntu.

Author: Great point.

            If you are using `Compute Canada <https://www.computecanada.ca/>`_, you can load modules as `mentioned here <https://intranet.neuro.polymtl.ca/computing-resources/compute-canada#modules>`_ and `also here <https://docs.computecanada.ca/wiki/Utiliser_des_modules/en#Loading_modules_automatically>`_.

Step 2: Install ``ivadomed``
++++++++++++++++++++++++++++

    .. tabs::

        .. tab:: PyPI Installation

            Install ``ivadomed`` and its requirements from
            `PyPI <https://pypi.org/project/ivadomed/>`__:

            ::

                pip install --upgrade pip

                pip install ivadomed

        .. tab:: Repo Installation (Advanced or Developer)

            Bleeding-edge developments are available on the project's master branch
            on Github. Install ``ivadomed`` from source:

            ::

                git clone https://github.com/ivadomed/ivadomed.git

                cd ivadomed

                pip install -e .

Step 3: Install ``torch`` and ``torchvision`` with CPU or GPU Support
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++

    .. tabs::

        .. tab:: CPU Support

>             If you plan to run ``ivadomed`` on CPU only, install pytorch per instructions provided below for your specific operating system:

Reviewer: Nitpick: `pytorch` -> `PyTorch`

```suggestion
            If you plan to run ``ivadomed`` on CPU only, install PyTorch per instructions provided below for your specific operating system:
```
